layout/scale.go

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Unlicense OR MIT
+
+package layout
+
+import (
+	"gioui.org/f32"
+	"gioui.org/op"
+	"gioui.org/op/clip"
+)
+
+// Fit scales the content to fit to the constraints.
+type Fit uint8
+
+const (
+	Contain Fit = iota
+	Cover
+	Fill
+	ScaleDown
+)
+
+func (fit Fit) Layout(gtx Context, w Widget) Dimensions {
+	macro := op.Record(gtx.Ops)
+	dims := w(gtx)
+	call := macro.Stop()
+
+	if dims.Size.X == 0 || dims.Size.Y == 0 {
+		call.Add(gtx.Ops)
+		return dims
+	}
+
+	scale := f32.Point{
+		X: float32(gtx.Constraints.Max.X) / float32(dims.Size.X),
+		Y: float32(gtx.Constraints.Max.Y) / float32(dims.Size.Y),
+	}
+
+	switch fit {
+	case Contain:
+		if scale.Y < scale.X {
+			scale.X = scale.Y
+		} else {
+			scale.Y = scale.X
+		}
+	case Cover:
+		if scale.Y > scale.X {
+			scale.X = scale.Y
+		} else {
+			scale.Y = scale.X
+		}
+	case ScaleDown:
+		if scale.Y < scale.X {
+			scale.X = scale.Y
+		} else {
+			scale.Y = scale.X
+		}
+		if scale.X >= 1 {
+			call.Add(gtx.Ops)
+			return dims
+		}
+	case Fill:
+	}
+
+	dims.Size.X = int(float32(dims.Size.X) * scale.X)
+	dims.Size.Y = int(float32(dims.Size.Y) * scale.Y)
```

Comment:
Add 0.5 to properly round sizes? Same goes for Baseline.

Reply:
I currently added this, however I'm not quite certain about it. I suspect this could end up causing a "half-pixel" gap in some scenarios.

Reply:
Good point, leave it out. Dimensions shouldn't be larger than the scaled size, even half a pixel.